x-pack/elastic-agent/pkg/composable/providers/kubernetes/node.go

```diff
@@ -0,0 +1,209 @@
+	if node.GetObjectMeta().GetDeletionTimestamp() != nil {
+		n.logger.Debugf("Watcher Node update (terminating): %+v", obj)
+		// Node is terminating, don't reload its configuration and ignore the event as long as node is Ready.
+		if isNodeReady(node) {
+			return
+		}
+		time.AfterFunc(n.cleanupTimeout, func() { n.emitStopped(node) })
+	} else {
+		n.logger.Debugf("Watcher Node update: %+v", obj)
+		n.emitStopped(node)
+		n.emitRunning(node)
+	}
+}
+
+// OnDelete ensures processing of node objects that are deleted
+func (n *node) OnDelete(obj interface{}) {
+	n.logger.Debugf("Watcher Node delete: %+v", obj)
+	node := obj.(*kubernetes.Node)
+	time.AfterFunc(n.cleanupTimeout, func() { n.emitStopped(node) })
+}
+
+func isUpdated(o, n interface{}) bool {
+	old, _ := o.(*kubernetes.Node)
+	new, _ := n.(*kubernetes.Node)
+
+	// Consider as not update in case one of the two objects is not a Node
+	if old == nil || new == nil {
+		return true
+	}
+
+	// This is a resync. It is not an update
+	if old.ResourceVersion == new.ResourceVersion {
+		return false
+	}
+
+	// If the old object and new object are different
+	oldCopy := old.DeepCopy()
+	oldCopy.ResourceVersion = ""
+
+	newCopy := new.DeepCopy()
+	newCopy.ResourceVersion = ""
+
+	// If the old object and new object are different in either meta or spec then there is a valid change
+	if !equality.Semantic.DeepEqual(oldCopy.Spec, newCopy.Spec) || !equality.Semantic.DeepEqual(oldCopy.ObjectMeta, newCopy.ObjectMeta) {
+		return true
+	}
+
+	// If there is a change in the node status then there is a valid change.
+	if isNodeReady(old) != isNodeReady(new) {
+		return true
+	}
+	return false
+}
+
+func getAddress(node *kubernetes.Node) string {
+	for _, address := range node.Status.Addresses {
+		if address.Type == v1.NodeExternalIP && address.Address != "" {
+			return address.Address
+		}
+	}
+
+	for _, address := range node.Status.Addresses {
+		if address.Type == v1.NodeInternalIP && address.Address != "" {
+			return address.Address
+		}
+	}
+
+	for _, address := range node.Status.Addresses {
+		if address.Type == v1.NodeHostName && address.Address != "" {
+			return address.Address
+		}
+	}
+
+	return ""
+}
+
+func isNodeReady(node *kubernetes.Node) bool {
+	for _, c := range node.Status.Conditions {
+		if c.Type == v1.NodeReady {
+			return c.Status == v1.ConditionTrue
+		}
+	}
+	return false
+}
+
+func generateNodeData(node *kubernetes.Node) *nodeData {
+	host := getAddress(node)
+
+	// If a node doesn't have an IP then dont monitor it
+	if host == "" {
+		return nil
+	}
+
+	// If the node is not in ready state then dont monitor it
+	if !isNodeReady(node) {
+		return nil
+	}
+
+	//TODO: add metadata here too ie -> meta := n.metagen.Generate(node)
+
+	// Pass annotations to all events so that it can be used in templating and by annotation builders.
+	annotations := common.MapStr{}
+	for k, v := range node.GetObjectMeta().GetAnnotations() {
+		safemapstr.Put(annotations, k, v)
```

Comment:
Should we be dedotting these?

Reply:
I was thinking about adding this when we will deal with metadata in general, but it's ok to add it now. Adding.